### ISSUE TITLE
Format nested keywords that run other keywords

### ISSUE THREAD

**Issue by Zeckie** [MarketSquare/robotframework-tidy#277]
Some of our tests use combinations of multiple built in `Run Keyword*` keywords ([documentation here](https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Run%20Keyword)) to handle or ignore errors, using indentation to make it easy to read. I ran robotidy on it, but robotidy removed the indentation, making it harder to read. Robotidy also does not seem to recognise which arguments are keywords and fix up their case.

I think robotidy should:
- understand at least the syntax of these built in keywords, such as which arguments are keywords or special values (like `AND`, `ELSE`, `ELSE IF`)
- indent nested keywords (each keyword should be on a new line)
- If possible, work with any keyword that is configured as a run keyword variant (`robot.Libraries.BuiltIn.run_keyword_variant` ?)

In this example, the keywords are `Run Keyword And Continue On Failure`, `Run Keyword If`, `Run Keywords`, and  `Log`. The only strings are `foo`, `bar` and `baz`.

Before:
```robot
*** Test Cases ***
Demo
    Run Keyword And Continue On Failure
    ...    Run Keyword If    ${True}
    ...      Run keywords
    ...        log    foo    AND
    ...        Log    bar
    ...    ELSE
    ...      log    baz
```

After (with existing robotframework-tidy 1.6.2):
```robot
*** Test Cases ***
Demo
    Run Keyword And Continue On Failure
    ...    Run Keyword If    ${True}
    ...    run keywords
    ...    log    foo    AND
    ...    Log    bar
    ...    ELSE
    ...    log    baz
```

Proposed:
```robot
*** Test Cases ***
Demo
    Run Keyword And Continue On Failure
    ...    Run Keyword If    ${True}
    ...        Run Keywords
    ...            Log    foo    AND
    ...            Log    bar
    ...    ELSE
    ...        Log    baz
```

from my pyproject.toml file:
```
[tool.robotidy]
configure = ["AlignTestCases:enabled=True", "RenameKeywords:enabled=True"]
```

**Comment by bhirsz** [comment]
It sounds good to me - Robotidy should definietly recognize run keyword variants (to avoid what happens in your example). Most likely it would be a separate transformer that could be easily configured/disabled if there is a need. 

First, you probably can disable NormalizeSeparator transformer and it should stop robotidy from messing with your keywords separators. When I will implement #267 it should also allow to disable formatting in given areas. Of course it's only temporary workaround.

1) I agree. Though how special arguments like AND, ELSE should be handled? I see it's handled in two ways in your example -> ELSE is split to a new line but on the same level as its parent keyword where AND stays in the line. How about "split to multiline, run keyword variants and their special arguments on the same indent level, the rest + 2 spaces"? Example:
```
*** Test Cases ***
Demo
    Run Keyword And Continue On Failure
    ...    Run Keyword If    ${True}
    ...        Run Keywords
    ...            Log    foo
    ...         AND
    ...            Log    bar
    ...    ELSE
    ...        Log    baz
```
Otherwise it could be like in your example if we decide that "ELSE, ELSE IF should be on a new line, same indent. AND should be in the same line as its parent keyword". 

2) Yes. It would work similarly how arguments in keywords in ``*** Settings ***`` section work (see example [here](https://github.com/MarketSquare/robotframework-tidy/blob/main/tests/atest/transformers/AlignSettingsSection/expected/multiline_keywords.robot)). The indent level would be the same as separator length parameter but we could overwrite it only for this parameter (ie. 4 spaces for all, but 2 spaces for this transformer for more compact code).
Of course special care should be taken for run keyword variants inside settings (ie Suite Setup, how it should be transformed with AlignSettingsSection? So it would not break)

3) I've never used this feature so I'm not sure but isn't it only available when I'm running the robot context? (if you registered keyword in your python code). That is not possible for us then. We can only use constant values. Hovewer it's certainly possible to allow registering variants using transformer configuration. For example (names are temporary):
```
robotidy -c IndentRunKeyword:run_keyword_variant=YourKeyword,SomeOtherKeyword
```

**Comment by Zeckie** [comment]
I don't have a strong opinion about the position of `AND`. I don't see the need for it to stand out like `ELSE`, and in most languages, it would have been a newline or semicolon. Having it stand out (on its own line) might be useful if the individual keywords under it are long enough to need multiple lines.

If the `Run Keyword And Continue On Failure` was not needed, it could be rewritten like this, without any need for the `AND`:
```
*** Test Cases ***
Demo
    IF    ${True}
        Log    foo
        Log    bar
    ELSE
        Log    baz
    END
```

It is probably not worth putting effort into getting this to work with custom run keyword variants, unless we know for sure that someone is going to use it. It doesn't look like they were ever really supported.

I found 2 more builtin keywords that run other keywords: `Wait Until Keyword Succeeds` and `Repeat Keyword`.

**Comment by bhirsz** [comment]
I've started working on this. I have working POC already for nested run keyword variants indentation. Currently the goal is to take care of possible branches (Else If, Else). It should be part of the 2nd release from now on (2.5, in about a week of anything goes according to plans).

Regarding your example from the first post - Run Keyword And Continue On Failure can be now rewritten with try except blocks. But there is still for run keywords variants that's why I will implement this issue. Most likely it will only support keyword calls in keywords and tests sections. Support for keywords in settings section and setting statements will come in further release.

**Comment by bhirsz** [comment]
The development is advanced hovewer I have few edge cases to cover. First one is comments. Iit should be easy - if any comment found inside keyword call, just "pop" it to the top:

```
Run Keywords
    Keyword
    Keyword  # comment
```

to
```
# comment
Run Keywords
    Keyword
    Keyword
```
It's not ideal but it's the easiest method of dealing with comments and not losing them. I can add something better in the future.

The second problem is with an AND. I've tried to keep it with the same line as the last keyword:
```
    Run Keywords
    ...    Log  1    AND
    ...    Log  2
```
But it doesn't look that good with more complex cases (it's not clear that AND refer to Run Keywords):
```
    Run Keywords
    ...    Run Keyword If  ${True}
    ...       Log  1  AND
    ...    Run Keyword If  ${True}
    ...        Log  2
```
That's why I will most likely put this on the same level as other keywords, just on new line:
```

```
    Run Keywords
    ...    Log  1
    ...    AND
    ...    Log  2

    Run Keywords
    ...    Run Keyword If  ${True}
    ...       Log  1
    ...    AND
    ...    Run Keyword If  ${True}
    ...        Log  2

```

**Comment by bhirsz** [comment]
Now I'm only wondering if we should indent AND keywords:
```
    Run Keywords
    ...    Log  1
    ...    AND
    ...    Log  2
```
vs
```
    Run Keywords
    ...        Log  1
    ...    AND
    ...        Log  2
```

**Comment by Zeckie** [comment]
I'm thinking the second option. It would make sense to indent the keywords (in this example the `Log`s) that are being separated by the `AND`.

**Comment by mnojek** [comment]
Well, for me it looks strange if keywords (`Log`s) are indented more than `AND` since they are all at the same level of abstraction. `AND` here just connects two arguments, which are keywords in this case. Yes, maybe it looks a bit more clear what's going on if we go with indented keywords, but why we would indent twice as deep when we start providing arguments for a keyword?

**Comment by bhirsz** [comment]
Two different opinions ;) And since I'm conflicted I think I will go with no intendation by default (for more concise output) but I will allow to be configured as an option. So everyone will be able to have the output they prefer. Something along:

robotidy -c IndentRunKeywords:indent_and_block=True .

(names are temporary).

**Comment by bhirsz** [comment]
Released in Robotidy 2.5.0.
